search/searcher/search_term.go

```diff
@@ -60,17 +64,82 @@ func NewTermSearcherBytes(ctx context.Context, indexReader index.IndexReader, te
 	if err != nil {
 		return nil, err
 	}
-	return newTermSearcherFromReader(indexReader, reader, term, field, boost, options)
+	return newTermSearcherFromReader(ctx, indexReader, reader, term, field, boost, options)
 }
 
-func newTermSearcherFromReader(indexReader index.IndexReader, reader index.TermFieldReader,
-	term []byte, field string, boost float64, options search.SearcherOptions) (*TermSearcher, error) {
+func tfTDFScoreMetrics(indexReader index.IndexReader) (uint64, float64, error) {
+	// default tf-idf stats
 	count, err := indexReader.DocCount()
+	if err != nil {
+		return 0, 0, err
+	}
+	fieldCardinality := 0
+
+	if count == 0 {
+		return 0, 0, nil
+	}
+	return count, float64(fieldCardinality / int(count)), nil
+}
+
+func bm25ScoreMetrics(ctx context.Context, field string,
+	indexReader index.IndexReader) (uint64, float64, error) {
+	var count uint64
+	var fieldCardinality int
+	var err error
+
+	bm25Stats, ok := ctx.Value(search.BM25PreSearchDataKey).(*search.BM25Stats)
+	if !ok {
+		count, err = indexReader.DocCount()
+		if err != nil {
+			return 0, 0, err
+		}
+		dict, err := indexReader.FieldDict(field)
+		if err != nil {
+			return 0, 0, err
+		}
+		fieldCardinality = dict.Cardinality()
+	} else {
+		count = uint64(bm25Stats.DocCount)
+		fieldCardinality, ok = bm25Stats.FieldCardinality[field]
+		if !ok {
+			return 0, 0, fmt.Errorf("field stat for bm25 not present %s", field)
+		}
+	}
+
+	if count == 0 && fieldCardinality == 0 {
+		return 0, 0, nil
+	}
+	return count, math.Ceil(float64(fieldCardinality) / float64(count)), nil
+}
+
+func newTermSearcherFromReader(ctx context.Context, indexReader index.IndexReader,
+	reader index.TermFieldReader, term []byte, field string, boost float64,
+	options search.SearcherOptions) (*TermSearcher, error) {
+	var count uint64
+	var avgDocLength float64
+	var err error
+
+	// as a fallback case we track certain stats for tf-idf scoring
+	count, avgDocLength, err = tfTDFScoreMetrics(indexReader)
 	if err != nil {
 		_ = reader.Close()
 		return nil, err
 	}
-	scorer := scorer.NewTermQueryScorer(term, field, boost, count, reader.Count(), options)
+	if ctx != nil {
+		if similaritModelCallback, ok := ctx.Value(search.
+			GetScoringModelCallbackKey).(search.GetScoringModelCallbackFn); ok {
+			similarityModel := similaritModelCallback()
+			if similarityModel == index.BM25Scoring {
+				// in case of bm25 need to fetch the multipliers as well (perhaps via context's presearch data)
+				count, avgDocLength, err = bm25ScoreMetrics(ctx, field, indexReader)
```

Comment:
Hmm it seems we'd be double calculating in the event of `bm25` scoring choice - perhaps an if-and-else is warranted here to either calculate just `tfTDFScoreMetrics` or `bm25ScoreMetrics`.

Reply:
👍🏼